### PR TITLE
GCP Observability: Add plugin registry API

### PR DESCRIPTION
**Pull request by yashykt**
Just adds the scaffolding for now in interest of doing it in a piece-wise manner. Proves that the integration with the stackdriver exporter works after https://github.com/grpc/grpc/pull/30567

### CODE REVIEW

File: include/grpcpp/opencensus.h [hunk 1]
```diff
@@ -42,6 +42,13 @@ void RegisterOpenCensusViewsForExport();
 // Returns the tracing Span for the current RPC.
 ::opencensus::trace::Span GetSpanFromServerContext(ServerContext* context);
 
+namespace experimental {
+// Setup gRPC's OpenCensus plugin to export monitoring and trace information to
+// stackdriver. Users of this API should not call `RegisterOpenCensusPlugin()`
+// or `RegisterOpenCensusViewsForExport()` in addition to this call.
+void RegisterOpenCensusPluginForGcpObservability();
```
Comment (sanjaypujare) [inline include/grpcpp/opencensus.h:49]: Is this what you will expose to the end users? In Java (and Go) we just have something like `grpcInit()` that users call which initializes everything for o11y. Will your `grpcInit` be implemented later?

File: BUILD [hunk 2]
```diff
@@ -7247,6 +7247,28 @@ grpc_cc_library(
     ],
 )
 
+# This is an EXPERIMENTAL target subject to change.
+grpc_cc_library(
+    name = "grpcpp_opencensus_gcp_observability_plugin",
```
Comment (sanjaypujare) [inline BUILD:7252]: I know we have called this observability code as plugin in various places but it is not really based on any plugin arch nor is it pluggable AFAICS. Again the name as opencensus in it whereas I thought we are going to hide that

Comment (sanjaypujare) [reply BUILD:7252]: As discussed offline: the goal of observability is to hide opencensus (or OTel in future) as implementation details but the C++ app developer is only aware of the grpc observability API.

File: src/cpp/ext/filters/census/gcp_observability.cc [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+constexpr uint32_t kMaxAnnotations = 128;
+constexpr uint32_t kMaxMessageEvents = 128;
+constexpr uint32_t kMaxLinks = 128;
+}  // namespace
+
+void RegisterOpenCensusPluginForGcpObservability() {
+  // TODO(yashykt): Add code for gRPC config parsing
+  grpc::RegisterOpenCensusPlugin();
+  grpc::RegisterOpenCensusViewsForExport();
+  // TODO(yashykt: Get probability from config
```
Comment (sanjaypujare) [inline src/cpp/ext/filters/census/gcp_observability.cc:41]: And also whether to enable/disable traces.

File: src/cpp/ext/filters/census/gcp_observability.cc [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  opencensus::trace::TraceConfig::SetCurrentTraceParams(
+      {kMaxAttributes, kMaxAnnotations, kMaxMessageEvents, kMaxLinks,
+       opencensus::trace::ProbabilitySampler(1.0)});
+  opencensus::exporters::trace::StackdriverOptions trace_opts;
+  // TODO(yashykt): Set up project ID based on config
+  opencensus::exporters::trace::StackdriverExporter::Register(
+      std::move(trace_opts));
+  opencensus::exporters::stats::StackdriverOptions stats_opts;
+  // TODO(yashykt): Set up project ID based on config
```
Comment (sanjaypujare) [inline src/cpp/ext/filters/census/gcp_observability.cc:50]: And also only if stats (metrics) is enabled

File: src/cpp/ext/filters/census/gcp_observability.cc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// TODO(yashykt): These constants are currently derived from the example at
+// https://cloud.google.com/traffic-director/docs/observability-proxyless#c++.
+// We might want these to be configurable.
+constexpr uint32_t kMaxAttributes = 128;
+constexpr uint32_t kMaxAnnotations = 128;
+constexpr uint32_t kMaxMessageEvents = 128;
+constexpr uint32_t kMaxLinks = 128;
+}  // namespace
+
+void RegisterOpenCensusPluginForGcpObservability() {
```
Comment (sanjaypujare) [inline src/cpp/ext/filters/census/gcp_observability.cc:37]: This should be named as `GcpObservability_grpcInit` or such

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
Added comments requesting changes

File: include/grpcpp/opencensus.h (hunk 1 above)
Comment (yashykt) [reply include/grpcpp/opencensus.h]: Moved to `GcpObservabiltyInit`

File: BUILD (hunk 2 above)
Comment (yashykt) [reply BUILD]: Renamed to `grpcpp_gcp_observability`

File: src/cpp/ext/filters/census/gcp_observability.cc (hunk 5 above)
Comment (yashykt) [reply src/cpp/ext/filters/census/gcp_observability.cc]: Done

File: src/cpp/ext/filters/census/gcp_observability.cc (hunk 3 above)
Comment (yashykt) [reply src/cpp/ext/filters/census/gcp_observability.cc]: Added TODO

File: src/cpp/ext/filters/census/gcp_observability.cc (hunk 4 above)
Comment (yashykt) [reply src/cpp/ext/filters/census/gcp_observability.cc]: Added TODO

File: src/cpp/ext/gcp_observability/gcp_observability.cc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <grpcpp/opencensus.h>
+
+namespace grpc {
+namespace experimental {
+
+namespace {
+// TODO(yashykt): These constants are currently derived from the example at
+// https://cloud.google.com/traffic-director/docs/observability-proxyless#c++.
+// We might want these to be configurable.
```
Comment (sanjaypujare) [inline src/cpp/ext/gcp_observability/gcp_observability.cc:38]: @DNVindhya we don't configure these in Java right? Why are there max values for attrs, annotations, events etc? Do we set them to the same values i.e. 128 for all 3?

**Review by sanjaypujare — APPROVED** [review]
LGTM (with one clarifying question added)

File: src/cpp/ext/gcp_observability/gcp_observability.cc (hunk 6 above)
Comment (DNVindhya) [reply src/cpp/ext/gcp_observability/gcp_observability.cc:38]: Yes, in Java we only configure the trace sampler based on Observability Config and for rest we use [default](https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/trace/config/TraceParams.java#L50) values.

Comment (sanjaypujare) [reply src/cpp/ext/gcp_observability/gcp_observability.cc:38]: So only events value is 128 and the other 2 are 32 each. Ideally we should have the same default values for all languages but right now it's not a blocker for this PR

**Pull request merged**